Speed Improvements in GAAClusterer

angeloskath opened the pull request:
I have added a similarity matrix to avoid recomputing the similarities. In addition the matrix update is done in a vectorized fashion to get the maximum NumPy has to offer.

The only thing that has been touched is the function cluster_vectorspace.

The result is more than 100 times faster clustering. It is an improvement (by better utilizing NumPy) from what is [mentioned to the nltk-dev group](https://groups.google.com/d/msg/nltk-dev/7D2ZTePSGFs/A83wLEm3AWIJ) as "First take on the problem".

angeloskath committed:
Speed improvement on GAAClusterer

New improved vectorized version of GAAClusterer that is orders of magnitude
faster than the previous one.

angeloskath committed:
Updates to GAAClusterer so that the new cluster_vectorspace function
conforms to PEP 8.

jnothman commented on nltk/cluster/gaac.py:
```diff
@@ -42,35 +41,52 @@ def cluster(self, vectors, assign_clusters=False, trace=False):
+            cl = cli+clj
+
+            # update for x<i
+            sim[:i, i] = (sim[:i, i]*cli + sim[:i, j]*clj)/cl
```
This would be easier to read, and more extensible to non-average link approaches if you factored out `(x * x_weight + y * y_weight) / (x_weight + y_weight)` as a separate function... or just use `np.average` for now.

angeloskath replied:
Renaming cli, clj, cl to i_weight, j_weight, weight_sum. About factoring it out as a function, I know it would be more extensible, I have opened a discussion in [nltk-dev](https://groups.google.com/forum/?fromgroups=#!forum/nltk-dev) regarding what should be added in general to hierarchical clustering in NLTK. Since the name of the class is GAAClusterer (Group Average Agglomerative) I don't really see the reason to factor it out as a function.

I do see the reason to factor out self._average_similarity though. Maybe we should discuss it in nltk-dev.

jnothman replied:
I suggest factoring it out not for extensibility but for readability: newcomers to clustering will be able to see quickly that these lines are doing the same thing as each other (and you will be assured there are no errors), and that thing will have a name.

jnothman replied:
And currently you're using `average_similarity` as if it were `similarity` and not average at all!

angeloskath replied:
ok, good point. I will also remove average similarity and import cosine_distance from util. I was trying not to change many things for some reason.

What about naming the function? average_link?

jnothman replied:
Even if you use `np.average` for now it is more readable, but `merge_similarities` is really what you're doing.

jnothman commented on nltk/cluster/gaac.py:
```diff
@@ -42,35 +41,52 @@ def cluster(self, vectors, assign_clusters=False, trace=False):
+            if trace:
+                print("merging %d and %d" % (i, j))
+
+            # update similarities
```
make the comment a bit more explanatory: "the new cluster i merged from i and j adopts the average of i and j's similarity to each other cluster"... not that I'm sure that's clear.

angeloskath replied:
I will add the comment you just wrote above. Since you are not sure that this is clear I could add something even more verbose.

"Cluster i can be thought to have i_weight points each having a similarity with cluster x equal to sim[i,x]. Cluster j can be thought to have j_weight points each having a similarity with cluster x equal to sim[j,x]. Merging clusters i and j the new cluster will have i_weight points with similarity sim[i,x] and j_weight points with similarity sim[j,x] the average similarity of the new cluster with cluster x is (sim[i,x]_i_weight + sim[j,x]_j_weight) / (weight_sum)"

jnothman commented on nltk/cluster/gaac.py:
```diff
@@ -42,35 +41,52 @@ def cluster(self, vectors, assign_clusters=False, trace=False):
+
+            # update the index map to reflect the indexes if we
+            # had removed j
+            for x in xrange(j+1, N):
```
This can be vectorized if `index_map` is a numpy array.

angeloskath replied:
Done.

jnothman commented on nltk/cluster/gaac.py:
```diff
@@ -42,35 +41,52 @@ def cluster(self, vectors, assign_clusters=False, trace=False):
+            # had removed j
+            for x in xrange(j+1, N):
+                index_map[x] -= 1
+            index_map[j] = -1
```
I assume this is like nulling a dead pointer, but it's no safer setting it to -1 than leaving it as is, if it's to be used as an offset into an array. You could set it to `N` to ensure an `IndexError` if it is accidentally used somewhere...?

angeloskath replied:
Ensuring IndexError was the point, it was just not really thought out very well (my C background caught up). Changing it to N.

angeloskath committed:
Readability changes and comment additions to GAAClusterer

- Removed _average_similarity and imported cosine_distance
- Added _average_link to be clear about how the distance
  matrix is updated after merge
- Changed sim to dist since we are using distances instead
  of similarities now

angeloskath committed:
Renaming _average_link to _merge_similarities

stevenbird commented:
@jnothman does this look ok to you now?

jnothman commented:
Looks fine, as long as O(n^2) memory is a reasonable sacrifice for the speed increase.

angeloskath commented:
I don't think O(n^2) is really a sacrifice since the previous implementation was pretty unusable with many points. For instance by projecting numbers in tests it would take the slow implementation ~6.5 hours to cluster 2400 two dimensional points and ~30 MB of memory and it takes this implementation 1.5 minutes and 44 MB of memory.

stevenbird commented:
@angeloskath - I see - that's certainly a reasonable sacrifice. I'll go ahead and merge this then. Thanks @jnothman for your input.
